compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -325,6 +326,95 @@ GatherOp::reifyResultShapes(OpBuilder &b,
       .reifyResultShapes(b, reifiedReturnShapes);
 }
 
+namespace {
+struct ConvertGatherToExtract
+    : public OpRewritePattern<IREE::LinalgExt::GatherOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(IREE::LinalgExt::GatherOp gatherOp,
+                                PatternRewriter &rewriter) const override {
+    // TODO: support memref case.
+    if (!gatherOp.hasPureTensorSemantics()) {
+      return failure();
+    }
+
+    auto loc = gatherOp.getLoc();
+    ArrayRef<int64_t> indicesShape = gatherOp.getIndicesType().getShape();
+    ArrayRef<int64_t> batchShape =
+        indicesShape.take_front(gatherOp.getBatchRank());
+    if (!llvm::all_of(batchShape, [](int64_t size) { return size == 1; })) {
+      return failure();
+    }
+
+    // Get all `indexDepth` indices as scalars.
+    SmallVector<Value> indices(indicesShape.size(),
+                               rewriter.create<arith::ConstantIndexOp>(loc, 0));
+    SmallVector<OpFoldResult> offsets(gatherOp.getIndexDepth());
+    for (int64_t i = 0; i < gatherOp.getIndexDepth(); i++) {
+      indices.back() = rewriter.create<arith::ConstantIndexOp>(loc, i);
+      Value elem = rewriter.create<tensor::ExtractOp>(
+          loc, gatherOp.getIndices(), indices);
+      offsets[i] =
+          rewriter
+              .create<arith::IndexCastOp>(loc, rewriter.getIndexType(), elem)
+              .getResult();
+    }
+
+    applyPermutationToVector(offsets, gatherOp.getDimensionMap());
+    int64_t sourceRank = gatherOp.getSourceType().getRank();
+    offsets.resize(sourceRank, rewriter.getIndexAttr(0));
+
+    // Create the new `tensor.extract_slice`
+    SmallVector<OpFoldResult> strides(sourceRank, rewriter.getIndexAttr(1));
+    SmallVector<int64_t> resultShape(gatherOp.getIndexDepth(), 1);
+    SmallVector<OpFoldResult> sizes(gatherOp.getIndexDepth(),
+                                    rewriter.getIndexAttr(1));
+    for (int64_t i = gatherOp.getIndexDepth(); i < sourceRank; i++) {
+      sizes.push_back(
+          rewriter.createOrFold<tensor::DimOp>(loc, gatherOp.getSource(), i));
+      resultShape.push_back(gatherOp.getSourceType().getDimSize(i));
+    }
+    auto resultType =
+        cast<RankedTensorType>(gatherOp.getSourceType()).clone(resultShape);
+    auto sliceOp = rewriter.create<tensor::ExtractSliceOp>(
+        loc, resultType, gatherOp.getSource(), offsets, sizes, strides);
+
+    // `sliceOp` may differ from the expected result type by leading unit
+    // dimensions. Reshape so that the types match.
+    int64_t sliceRank = sliceOp.getResultType().getRank();
+    int64_t gatherRank = gatherOp.getOutputType().getRank();
+    if (sliceRank < gatherRank) {
+      SmallVector<ReassociationIndices> reassoc(1);
+      llvm::append_range(reassoc[0],
+                         llvm::seq<int64_t>(0, gatherOp.getBatchRank()));
```

Comment:
I recently learned that it can be `llvm::seq<int64_t>(gatherOp.getBatchRank())`.

https://github.com/llvm/llvm-project/blob/4042a002cea6dc6f12e32953c820f6eae1ac1817/llvm/include/llvm/ADT/Sequence.h#L309-L317

```cpp
/// Iterate over an integral type from 0 up to - but not including - Size.
/// Note: Size value has to be within [INTMAX_MIN, INTMAX_MAX - 1] for
/// forward iteration (resp. [INTMAX_MIN + 1, INTMAX_MAX - 1] for reverse
/// iteration).
template <typename T, typename = std::enable_if_t<std::is_integral<T>::value &&
                                                  !std::is_enum<T>::value>>
auto seq(T Size) {
  return seq<T>(0, Size);
}
```